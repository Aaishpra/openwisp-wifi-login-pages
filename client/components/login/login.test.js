/* eslint-disable prefer-promise-reject-errors */
import axios from "axios";
import {shallow} from "enzyme";
import React from "react";
import {BrowserRouter as Router} from "react-router-dom";
import renderer from "react-test-renderer";

import getConfig from "../../utils/get-config";
import Login from "./login";

jest.mock("axios");

const defaultConfig = getConfig("default");
const createTestProps = props => {
  return {
    language: "en",
    orgSlug: "default",
    loginForm: defaultConfig.components.login_form,
    privacyPolicy: defaultConfig.privacy_policy,
    termsAndConditions: defaultConfig.terms_and_conditions,
    authenticate: jest.fn(),
    ...props,
  };
};
describe("<Login /> rendering", () => {
  let props;
  it("should render correctly without social links", () => {
    props = createTestProps();
    const component = renderer
      .create(
        <Router>
          <Login {...props} />
        </Router>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
  it("should render correctly with social links", () => {
    props = createTestProps({
      loginForm: {
        ...defaultConfig.components.login_form,
        social_login: {
          ...defaultConfig.components.login_form,
          links: [
            {
              text: {
                en: "Facebook",
              },
              url: "test url",
              icon: null,
            },
            {
              icon: "test.png",
              url:
                "https://control.co.ke/accounts/facebook/login/?next=%2Ffreeradius%2Fsocial-login%2Fstaging%2F%3Fcp%3Dhttp%3A%2F%2Fcontrol.brandfi.co.ke%2Floginpage%2F%26last%3D",
            },
          ],
        },
      },
    });
    const component = renderer
      .create(
        <Router>
          <Login {...props} />
        </Router>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});

describe("<Login /> interactions", () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<Login {...props} />);
  });
  it("should change state values when handleChange function is invoked", () => {
    wrapper
      .find("#owisp-login-username")
      .simulate("change", {target: {value: "test username", name: "username"}});
    expect(wrapper.state("username")).toEqual("test username");
    wrapper
      .find("#owisp-login-password")
      .simulate("change", {target: {value: "test password", name: "password"}});
    expect(wrapper.state("password")).toEqual("test password");
  });
  it("should execute handleSubmit correctly when form is submitted", () => {
    axios
      .mockImplementationOnce(() => {
        return Promise.reject({
          response: {
            data: {
              username: "username error",
              password: "password error",
              detail: "error details",
              non_field_errors: "non field errors",
            },
          },
        });
      })
      .mockImplementationOnce(() => {
        return Promise.resolve();
      });
    const event = {preventDefault: () => {}};
    return wrapper
      .instance()
      .handleSubmit(event)
      .then(() => {
        expect(wrapper.instance().state.errors).toEqual({
          username: "username error",
          nonField: "error details",
          password: "password error",
        });
        expect(wrapper.find(".owisp-login-error")).toHaveLength(3);
      })
      .then(() => {
        return wrapper
          .instance()
          .handleSubmit(event)
          .then(() => {
            expect(wrapper.instance().state.errors).toEqual({});
            expect(
              wrapper.instance().props.authenticate.mock.calls.length,
            ).toBe(1);
          });
      });
  });
});
