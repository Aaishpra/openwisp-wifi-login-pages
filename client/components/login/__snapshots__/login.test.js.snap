// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> rendering should render correctly with social links 1`] = `
<div
  className="owisp-login-container"
>
  <form
    className="owisp-login-form"
    onSubmit={[Function]}
  >
    <div
      className="owisp-login-header"
    >
      <div
        className="owisp-login-header-content"
      >
         sign in
      </div>
    </div>
    <div
      className="owisp-login-fieldset"
    >
      <label
        className="owisp-login-label owisp-login-label-username"
        htmlFor="owisp-login-username"
      >
        <div
          className="owisp-login-label-text"
        >
          username
        </div>
        <input
          className="owisp-login-input
                      owisp-login-input-username
                      "
          id="owisp-login-username"
          name="username"
          onChange={[Function]}
          pattern="[a-zA-Z@.+\\\\-_\\\\d]{1,150}"
          placeholder="enter username"
          required={true}
          title="only letters, numbers, and @/./+/-/_ characters"
          type="text"
          value=""
        />
      </label>
      <label
        className="owisp-login-label owisp-login-label-password"
        htmlFor="owisp-login-password"
      >
        <div
          className="owisp-login-label-text"
        >
          password
        </div>
        <input
          className="owisp-login-input owisp-login-input-password
                      "
          id="owisp-login-password"
          name="password"
          onChange={[Function]}
          pattern=".{6,}"
          placeholder="password"
          required={true}
          title="password must be a minimum of 6 characters"
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      className="owisp-login-add-info"
    >
      By logging in, you accept the 
      <a
        className="owisp-login-additional-link"
        href="/default/login/terms-and-conditions"
        onClick={[Function]}
      >
        terms and conditions
      </a>
       and the 
      <a
        className="owisp-login-additional-link"
        href="/default/login/privacy-policy"
        onClick={[Function]}
      >
        privacy policy
      </a>
       of this WiFi service. 
    </div>
    <input
      className="owisp-login-form-btn owisp-login-login-btn"
      id="owisp-login-login-btn"
      type="submit"
      value="sign in"
    />
    <div
      className="owisp-login-social-links-div"
    >
      <a
        className="owisp-login-social-link"
        href="test url"
        rel="noopener noreferrer"
        target="_blank"
      >
        <div
          className="owisp-login-social-link-text"
        >
          Facebook
        </div>
      </a>
      <a
        className="owisp-login-social-link"
        href="https://control.co.ke/accounts/facebook/login/?next=%2Ffreeradius%2Fsocial-login%2Fstaging%2F%3Fcp%3Dhttp%3A%2F%2Fcontrol.brandfi.co.ke%2Floginpage%2F%26last%3D"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="https://control.co.ke/accounts/facebook/login/?next=%2Ffreeradius%2Fsocial-login%2Fstaging%2F%3Fcp%3Dhttp%3A%2F%2Fcontrol.brandfi.co.ke%2Floginpage%2F%26last%3D"
          className="owisp-login-social-link-icon"
          src="/assets/default/test.png"
        />
      </a>
    </div>
    <div
      className="owisp-login-links-div"
    >
      <a
        className="owisp-login-link"
        href="/default/password/reset"
        onClick={[Function]}
      >
        Forgot your password?
      </a>
      <a
        className="owisp-login-link"
        href="/default/registration"
        onClick={[Function]}
      >
        Not registered yet? Create an account
      </a>
    </div>
  </form>
</div>
`;

exports[`<Login /> rendering should render correctly without social links 1`] = `
<div
  className="owisp-login-container"
>
  <form
    className="owisp-login-form"
    onSubmit={[Function]}
  >
    <div
      className="owisp-login-header"
    >
      <div
        className="owisp-login-header-content"
      >
         sign in
      </div>
    </div>
    <div
      className="owisp-login-fieldset"
    >
      <label
        className="owisp-login-label owisp-login-label-username"
        htmlFor="owisp-login-username"
      >
        <div
          className="owisp-login-label-text"
        >
          username
        </div>
        <input
          className="owisp-login-input
                      owisp-login-input-username
                      "
          id="owisp-login-username"
          name="username"
          onChange={[Function]}
          pattern="[a-zA-Z@.+\\\\-_\\\\d]{1,150}"
          placeholder="enter username"
          required={true}
          title="only letters, numbers, and @/./+/-/_ characters"
          type="text"
          value=""
        />
      </label>
      <label
        className="owisp-login-label owisp-login-label-password"
        htmlFor="owisp-login-password"
      >
        <div
          className="owisp-login-label-text"
        >
          password
        </div>
        <input
          className="owisp-login-input owisp-login-input-password
                      "
          id="owisp-login-password"
          name="password"
          onChange={[Function]}
          pattern=".{6,}"
          placeholder="password"
          required={true}
          title="password must be a minimum of 6 characters"
          type="password"
          value=""
        />
      </label>
    </div>
    <div
      className="owisp-login-add-info"
    >
      By logging in, you accept the 
      <a
        className="owisp-login-additional-link"
        href="/default/login/terms-and-conditions"
        onClick={[Function]}
      >
        terms and conditions
      </a>
       and the 
      <a
        className="owisp-login-additional-link"
        href="/default/login/privacy-policy"
        onClick={[Function]}
      >
        privacy policy
      </a>
       of this WiFi service. 
    </div>
    <input
      className="owisp-login-form-btn owisp-login-login-btn"
      id="owisp-login-login-btn"
      type="submit"
      value="sign in"
    />
    <hr
      className="owisp-login-social-hr"
      data-content="OR"
    />
    <div
      className="owisp-login-divider-description"
    >
      login with
    </div>
    <div
      className="owisp-login-social-links-div"
    />
    <div
      className="owisp-login-links-div"
    >
      <a
        className="owisp-login-link"
        href="/default/password/reset"
        onClick={[Function]}
      >
        Forgot your password?
      </a>
      <a
        className="owisp-login-link"
        href="/default/registration"
        onClick={[Function]}
      >
        Not registered yet? Create an account
      </a>
    </div>
  </form>
</div>
`;
